File: Ruby/searching/binary_search.rb
# Binary search implementation in Ruby
# Time Complexity: O(log n)
# Space Complexity: O(log n) for recursive, O(1) for iterative

class BinarySearch
  # Recursive binary search implementation
  # Time Complexity: O(log n), Space Complexity: O(log n)
  def self.recursive(arr, target, low = 0, high = nil)
    high = arr.length - 1 if high.nil?
    return nil if low > high
    
    mid = (low + high) / 2
    
    case arr[mid] <=> target
    when 0
      mid
    when 1
      recursive(arr, target, low, mid - 1)
    when -1
      recursive(arr, target, mid + 1, high)
    end
  end

  # Iterative binary search implementation
  # Time Complexity: O(log n), Space Complexity: O(1)
  def self.iterative(arr, target)
    low = 0
    high = arr.length - 1
    
    while low <= high
      mid = (low + high) / 2
      
      case arr[mid] <=> target
      when 0
        return mid
      when 1
        high = mid - 1
      when -1
        low = mid + 1
      end
    end
    
    nil
  end

  # Binary search that returns the leftmost occurrence
  # Time Complexity: O(log n), Space Complexity: O(1)
  def self.find_first(arr, target)
    low = 0
    high = arr.length - 1
    result = nil
    
    while low <= high
      mid = (low + high) / 2
      
      if arr[mid] == target
        result = mid
        high = mid - 1  # Continue searching left
      elsif arr[mid] < target
        low = mid + 1
      else
        high = mid - 1
      end
    end
    
    result
  end

  # Binary search that returns the rightmost occurrence
  # Time Complexity: O(log n), Space Complexity: O(1)
  def self.find_last(arr, target)
    low = 0
    high = arr.length - 1
    result = nil
    
    while low <= high
      mid = (low + high) / 2
      
      if arr[mid] == target
        result = mid
        low = mid + 1  # Continue searching right
      elsif arr[mid] < target
        low = mid + 1
      else
        high = mid - 1
      end
    end
    
    result
  end

  # Find insertion point for target in sorted array
  # Time Complexity: O(log n), Space Complexity: O(1)
  def self.insertion_point(arr, target)
    low = 0
    high = arr.length
    
    while low < high
      mid = (low + high) / 2
      
      if arr[mid] < target
        low = mid + 1
      else
        high = mid
      end
    end
    
    low
  end
end

# Example usage and testing
if __FILE__ == $0
  puts "Binary Search Examples:"
  
  sorted_list = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
  puts "Sorted array: #{sorted_list}"
  
  # Test basic searches
  target = 5
  puts "\nSearching for #{target}:"
  puts "Recursive: #{BinarySearch.recursive(sorted_list, target)}"
  puts "Iterative: #{BinarySearch.iterative(sorted_list, target)}"
  
  # Test not found
  target = 11
  puts "\nSearching for #{target} (not in array):"
  puts "Recursive: #{BinarySearch.recursive(sorted_list, target)}"
  puts "Iterative: #{BinarySearch.iterative(sorted_list, target)}"
  
  # Test with duplicates
  arr_with_duplicates = [1, 2, 2, 2, 3, 4, 5, 5, 5, 6]
  target = 2
  puts "\nArray with duplicates: #{arr_with_duplicates}"
  puts "Searching for #{target}:"
  puts "Any occurrence: #{BinarySearch.iterative(arr_with_duplicates, target)}"
  puts "First occurrence: #{BinarySearch.find_first(arr_with_duplicates, target)}"
  puts "Last occurrence: #{BinarySearch.find_last(arr_with_duplicates, target)}"
  
  target = 5
  puts "\nSearching for #{target}:"
  puts "First occurrence: #{BinarySearch.find_first(arr_with_duplicates, target)}"
  puts "Last occurrence: #{BinarySearch.find_last(arr_with_duplicates, target)}"
  
  # Test insertion point
  puts "\nInsertion points:"
  [0, 2.5, 5, 7].each do |val|
    puts "Insertion point for #{val}: #{BinarySearch.insertion_point(sorted_list, val)}"
  end
end

File: Swift/searching/binarySearch.swift
/*
 * ## Problem: Binary Search ##
 *
 * Binary search is an efficient algorithm for finding an item from a **sorted** list of items.
 * It works by repeatedly dividing the search interval in half.
 *
 * How it works:
 * 1.  Compare the target value to the middle element of the array.
 * 2.  If they are not equal, the half in which the target cannot lie is eliminated,
 * and the search continues on the remaining half.
 * 3.  This process is repeated until the target value is found or the remaining half is empty.
 *
 * Prerequisite: The input array must be sorted.
 */

// An iterative implementation of Binary Search.
func binarySearchIterative(array: [Int], target: Int) -> Int? {
    // Set the initial search range.
    var low = 0
    var high = array.count - 1

    // Continue searching as long as the range is valid.
    while low <= high {
        // Find the middle index.
        // Using (low + (high - low) / 2) avoids potential overflow.
        let mid = low + (high - low) / 2

        // Check if the middle element is the target.
        if array[mid] == target {
            return mid // Target found, return its index.
        } else if array[mid] < target {
            // If the middle element is less than the target,
            // narrow the search to the right half.
            low = mid + 1
        } else {
            // If the middle element is greater than the target,
            // narrow the search to the left half.
            high = mid - 1
        }
    }
    
    // If the loop finishes, the target was not in the array.
    return nil
}

// A recursive implementation of Binary Search.
func binarySearchRecursive(array: [Int], target: Int, range: Range<Int>) -> Int? {
    // Base case: If the search range is empty, the element is not present.
    if range.lowerBound >= range.upperBound {
        return nil
    }

    // Find the middle index of the current range.
    let midIndex = range.lowerBound + (range.upperBound - range.lowerBound) / 2

    // Check if the middle element is the target.
    if array[midIndex] == target {
        return midIndex
    } else if array[midIndex] < target {
        // Search in the right half of the current range.
        return binarySearchRecursive(array: array, target: target, range: (midIndex + 1)..<range.upperBound)
    } else {
        // Search in the left half of the current range.
        return binarySearchRecursive(array: array, target: target, range: range.lowerBound..<midIndex)
    }
}


// --- Test Cases ---
let sortedArray = [2, 5, 8, 12, 16, 23, 38, 56, 72, 91]
let target1 = 23
let target2 = 99

print("Array: \(sortedArray)")

print("\n--- Iterative ---")
if let index = binarySearchIterative(array: sortedArray, target: target1) {
    print("Element \(target1) found at index \(index).")
} else {
    print("Element \(target1) not found in the array.")
}

if let index = binarySearchIterative(array: sortedArray, target: target2) {
    print("Element \(target2) found at index \(index).")
} else {
    print("Element \(target2) not found in the array.")
}


print("\n--- Recursive ---")
if let index = binarySearchRecursive(array: sortedArray, target: target1, range: 0..<sortedArray.count) {
    print("Element \(target1) found at index \(index).")
} else {
    print("Element \(target1) not found in the array.")
}

if let index = binarySearchRecursive(array: sortedArray, target: target2, range: 0..<sortedArray.count) {
    print("Element \(target2) found at index \(index).")
} else {
    print("Element \(target2) not found in the array.")
}
